composer/trainer/_deepspeed.py

```diff
@@ -75,25 +75,23 @@ def _ensure_no_optim_in_config(config: Dict[str, Any]):
 def _add_precision_config(config: Dict[str, Any], state: State):
     precision = state.precision
 
+    # Verify DeepSpeed config is consistent with state.precision if set
     ds_precision = None
     if 'fp16' in config and 'enabled' in config['fp16'] and config['fp16']['enabled']:
-        ds_precision = Precision.FP16
-    if 'bf16' in config and 'enabled' in config['bf16'] and config['bf16']['enabled']:
-        raise ValueError(('DeepSpeed is configured to use BFLOAT16, but this is unsupported by the '
-                          'Mosaic trainer.'))
-    if 'amp' in config and 'enabled' in config['amp'] and config['amp']['enabled']:
-        raise ValueError(('DeepSpeed is configured to use Apex AMP, but this is unsupported by the '
-                          'Mosaic trainer.'))
-
+        ds_precision = Precision.AMP_FP16
+    elif 'bf16' in config and 'enabled' in config['bf16'] and config['bf16']['enabled']:
+        ds_precision = Precision.AMP_BF16
+    elif 'amp' in config and 'enabled' in config['amp'] and config['amp']['enabled']:
```

Comment:
The Deepspeed naming conventions are from their docs / their config. It is not our convention.

Reply:
Why are 'amp' and 'FP16' the same under the hood? Shouldn't we say this to users? The deepspeed user expects 'amp' and FP16 to work differently, and we are not doing this? Do Deepspeed's FP16 and BF16 even correspond to AMP_FP16 and AMP_BF16? Or are they closer to pure FP16 and BF16 training?

Reply:
I updated comment to link to DeepSpeed docs. We do:
1) if user specifies AMP_FP16 or AMP_BF16 + deepspeed, we remap to amp or bf16 in deepspeed config which are the closest ones
2) if a user specifies precision in deepspeed config, we verify what they passed in to composer is approximately the same to ensure they didn't accidentally only change in composer

When running, only the deepspeed config is used -- the Composer precision is ignored.

fp16 corresponds to amp in deepspeed with optimization level o2 iirc. It is not pure fp16

Reply:
Okay. I'd favor a design which explicitly disallows specifying the composer precision when deepspeed is used and require a Deepspeed precision be passed in, so things are clear to users. Especially now that our precision names do not line up. But if this is how we are already operating, then okay.

Reply:
hm... I think from a usability perspective it is nice to just specify the composer one and have it happen automatically. In either case, since we aren't super actively maintaining deepspeed path, im going to keep with what ran before this PR